lib/event_sourcery/event_processing/esp_runner.rb

```diff
@@ -7,23 +7,28 @@ def initialize(event_processors:,
                      event_source:,
                      max_seconds_for_processes_to_terminate: 30,
                      shutdown_requested: false,
-                     after_fork: nil)
+                     after_fork: nil,
+                     after_subprocess_termination: nil,
+                     logger: EventSourcery.logger)
         @event_processors = event_processors
         @event_source = event_source
-        @pids = []
+        @pids = {}
         @max_seconds_for_processes_to_terminate = max_seconds_for_processes_to_terminate
         @shutdown_requested = shutdown_requested
         @exit_status = true
         @after_fork = after_fork
+        @after_subprocess_termination = after_subprocess_termination
+        @logger = logger
       end
 
       # Start each Event Stream Processor in a new child process.
       def start!
         with_logging do
           start_processes
           listen_for_shutdown_signals
-          wait_till_shutdown_requested
-          record_terminated_processes
+          while_waiting_for_shutdown do
```

Comment:
why have you chosen to supply a block to this method? 
In terms of reading the code it makes sense.
My only comment (I'm quite sure this is a nitpick) is that if this is the only block being passed to this method.. why can you just replace the `yield` with `record_terminated_processes`?

Reply:
humm to be honest I think it's ok like this.. I think it reads better.. you would have to change the method name to something like
`record_terminated_processes_while_waiting_for_shutdown` might be a little too verbose.

Reply:
Yeah, that'd work. I'm aiming to have the contents of this `start!` method to provide an understandable  overview. To do this I try to provide a consistent level of abstraction with the statements. Were we to replace `yield` with `record_terminated_processes`, I'd feel this should be communicated via the method name. So I'd probably come up with something like `record_terminated_processes_while_waiting_for_shutdown`. Which isn't too bad, but perhaps a bit long. That would be how I'd do it in a language that doesn't support blocks. I like how using the block it hints that `record_terminated_processes` is (could be) called multiple times. Being able to do this is one of the nice things about Ruby.

Reply:
I'm interested in your reaction though. Is it confusing? Convoluted?

Reply:
No certainly not confusing or convoluted. It was just a comment regarding if the concern of recording the terminated processes, if it can just belong to `while_waiting_for_shutdown` or not but I certainly think the way it is not is fine and reads well. all good 👍

Reply:
Cheers.